### ISSUE TITLE
okta-spring: some security filters are not present in Native image

### ISSUE THREAD

**Issue by ivangfr** [spring-projects-experimental/spring-native#1469]
Hi, I've updated my project [okta-springboot](https://github.com/ivangfr/okta-springboot) to **spring-boot** `2.6.3`, **spring-native** `0.11.2` and **okta-spring** `2.1.4`

After building successfully the Docker native image, we can see that some security filters present in JVM logs (marked with *) are not present in Native logs

- JVM
  ```
  INFO 1 --- [           main] o.s.s.web.DefaultSecurityFilterChain     : Will secure any request with [
    org.springframework.security.web.context.request.async.WebAsyncManagerIntegrationFilter@1682e6a,
    org.springframework.security.web.context.SecurityContextPersistenceFilter@25218a4d,
    org.springframework.security.web.header.HeaderWriterFilter@7cb29ea8,
    org.springframework.security.web.csrf.CsrfFilter@7b1a30e5,
    org.springframework.security.web.authentication.logout.LogoutFilter@64c8fcfb,
  * org.springframework.security.oauth2.client.web.OAuth2AuthorizationRequestRedirectFilter@7af0d91b,
  * org.springframework.security.oauth2.client.web.OAuth2LoginAuthenticationFilter@1c7f6e96,
  * org.springframework.security.web.authentication.ui.DefaultLoginPageGeneratingFilter@23f60b7d,
  * org.springframework.security.web.authentication.ui.DefaultLogoutPageGeneratingFilter@505d2bac,
    org.springframework.security.web.savedrequest.RequestCacheAwareFilter@56da96b3,
    org.springframework.security.web.servletapi.SecurityContextHolderAwareRequestFilter@59ec7020,
    org.springframework.security.web.authentication.AnonymousAuthenticationFilter@4cecc15a,
    org.springframework.security.web.session.SessionManagementFilter@7e747037,
    org.springframework.security.web.access.ExceptionTranslationFilter@2fca3eb5,
    org.springframework.security.web.access.intercept.FilterSecurityInterceptor@464ede1f
  ]
  ```

- Native
  ```
  INFO 1 --- [           main] o.s.s.web.DefaultSecurityFilterChain     : Will secure any request with [
    org.springframework.security.web.context.request.async.WebAsyncManagerIntegrationFilter@73703f0e,
    org.springframework.security.web.context.SecurityContextPersistenceFilter@74eb08bd,
    org.springframework.security.web.header.HeaderWriterFilter@7146881f,
    org.springframework.security.web.csrf.CsrfFilter@4ec5994c,
    org.springframework.security.web.authentication.logout.LogoutFilter@17f0c2ea,
    org.springframework.security.web.savedrequest.RequestCacheAwareFilter@6b94fe94,
    org.springframework.security.web.servletapi.SecurityContextHolderAwareRequestFilter@3221e5ad,
    org.springframework.security.web.authentication.AnonymousAuthenticationFilter@193869a3,
    org.springframework.security.web.session.SessionManagementFilter@6bc076fd,
    org.springframework.security.web.access.ExceptionTranslationFilter@4548022f,
    org.springframework.security.web.access.intercept.FilterSecurityInterceptor@2393f04b
  ]
  ```

**Comment by snicoll** [comment]
When you say JVM, do you mean JVM with AOT or plain JVM? If the filters are not present when AOT has processed the context, it could be that a property was not set at build time that triggered a condition to create those filters.

**Comment by ivangfr** [comment]
Hi @snicoll JVM with AOT mode disabled

**Comment by snicoll** [comment]
I can have a look to the sample when time permits but looking at the auto-configuration for the missing filters and the conditions associated could give you a clue of what the problem might be. Also, it's not clear if you were using Spring Native 0.11 before or if you upgraded from 0.10.

**Comment by ivangfr** [comment]
Honestly, for now, I have no idea what can be the problem. Also, it was working on the previous version of my project, i.e commit `3a9101ab897204aca6ea27ebb01a3c3cdbb6e4df`. In this case, it was using **spring-boot** `2.5.7`, **spring-native** `0.10.5` and **okta-spring** `2.1.3`.

I've just checkout `3a9101ab897204aca6ea27ebb01a3c3cdbb6e4df` (previous commit), built and ran the Docker native image. Logs below

```
➜  okta-springboot git:(3a9101a) docker run --rm --name simple-service -p 8080:8080 \
      -e OKTA_CLIENT_ID=${OKTA_CLIENT_ID} \
      -e OKTA_CLIENT_SECRET=${OKTA_CLIENT_SECRET} \
      -e OKTA_DOMAIN=${OKTA_DOMAIN} \
      ivanfranchin/simple-service:1.0.0
2022-01-27 12:08:57.601  INFO 1 --- [           main] o.s.nativex.NativeListener               : This application is bootstrapped with code generated with Spring AOT

  .   ____          _            __ _ _
 /\\ / ___'_ __ _ _(_)_ __  __ _ \ \ \ \
( ( )\___ | '_ | '_| | '_ \/ _` | \ \ \ \
 \\/  ___)| |_)| | | | | || (_| |  ) ) ) )
  '  |____| .__|_| |_|_| |_\__, | / / / /
 =========|_|==============|___/=/_/_/_/
 :: Spring Boot ::                (v2.5.7)

2022-01-27 12:08:57.604  INFO 1 --- [           main] c.m.s.SimpleServiceApplication           : Starting SimpleServiceApplication using Java 11.0.13 on 0ad2299b8798 with PID 1 (/workspace/com.mycompany.simpleservice.SimpleServiceApplication started by cnb in /workspace)
2022-01-27 12:08:57.604  INFO 1 --- [           main] c.m.s.SimpleServiceApplication           : No active profile set, falling back to default profiles: default
2022-01-27 12:08:57.755  INFO 1 --- [           main] o.s.b.w.embedded.tomcat.TomcatWebServer  : Tomcat initialized with port(s): 8080 (http)
2022-01-27 12:08:57.756  INFO 1 --- [           main] o.apache.catalina.core.StandardService   : Starting service [Tomcat]
2022-01-27 12:08:57.757  INFO 1 --- [           main] org.apache.catalina.core.StandardEngine  : Starting Servlet engine: [Apache Tomcat/9.0.55]
2022-01-27 12:08:57.764  INFO 1 --- [           main] o.a.c.c.C.[Tomcat].[localhost].[/]       : Initializing Spring embedded WebApplicationContext
2022-01-27 12:08:57.764  INFO 1 --- [           main] w.s.c.ServletWebServerApplicationContext : Root WebApplicationContext: initialization completed in 159 ms
2022-01-27 12:08:57.784  WARN 1 --- [           main] i.m.c.i.binder.jvm.JvmGcMetrics          : GC notifications will not be available because MemoryPoolMXBeans are not provided by the JVM
2022-01-27 12:08:59.370  INFO 1 --- [           main] o.s.s.web.DefaultSecurityFilterChain     : Will secure any request with [
        org.springframework.security.web.context.request.async.WebAsyncManagerIntegrationFilter@109907e4,
        org.springframework.security.web.context.SecurityContextPersistenceFilter@283ed47c,
        org.springframework.security.web.header.HeaderWriterFilter@6e856c7,
        org.springframework.security.web.csrf.CsrfFilter@629071b9,
        org.springframework.security.web.authentication.logout.LogoutFilter@42717d25,
        org.springframework.security.oauth2.client.web.OAuth2AuthorizationRequestRedirectFilter@29ec3be9,
        org.springframework.security.oauth2.client.web.OAuth2LoginAuthenticationFilter@1de1fb70,
        org.springframework.security.web.authentication.ui.DefaultLoginPageGeneratingFilter@134dd8ba,
        org.springframework.security.web.authentication.ui.DefaultLogoutPageGeneratingFilter@33143623,
        org.springframework.security.web.savedrequest.RequestCacheAwareFilter@561d4466,
        org.springframework.security.web.servletapi.SecurityContextHolderAwareRequestFilter@22a56975,
        org.springframework.security.web.authentication.AnonymousAuthenticationFilter@67ce5b4,
        org.springframework.security.web.session.SessionManagementFilter@39c17512,
        org.springframework.security.web.access.ExceptionTranslationFilter@701392da,
        org.springframework.security.web.access.intercept.FilterSecurityInterceptor@5c1e4790
]
2022-01-27 12:08:59.387  INFO 1 --- [           main] o.s.b.a.e.web.EndpointLinksResolver      : Exposing 6 endpoint(s) beneath base path '/actuator'
2022-01-27 12:08:59.401  INFO 1 --- [           main] o.s.b.w.embedded.tomcat.TomcatWebServer  : Tomcat started on port(s): 8080 (http) with context path ''
2022-01-27 12:08:59.403  INFO 1 --- [           main] c.m.s.SimpleServiceApplication           : Started SimpleServiceApplication in 1.683 seconds (JVM running for 1.685)
```

As we can see, the filters are there.

**Comment by snicoll** [comment]
```
 -e OKTA_CLIENT_ID=${OKTA_CLIENT_ID} \
 -e OKTA_CLIENT_SECRET=${OKTA_CLIENT_SECRET} \
 -e OKTA_DOMAIN=${OKTA_DOMAIN} \
```

Did you consider [my previous comment](https://github.com/spring-projects-experimental/spring-native/issues/1469#issuecomment-1022994212) and [this section of the doc](https://docs.spring.io/spring-native/docs/current/reference/htmlsingle/#aot-bean-factory-preparation)? One way to know for sure is running the app without any extra property and checking that the filters are present (so without setting those properties in your IDE or something).

**Comment by ivangfr** [comment]
Hi @snicoll thanks for remind me about you previous comment!

I've just hardcoded the values of those **ENV VARS** in `application.properties`.
```
...
okta.oauth2.issuer=https://...
okta.oauth2.client-id=0oaf...
okta.oauth2.client-secret=bKKL...
...
```

Same problem.

- Running JVM docker image with AOT mode disabled

  ```
  2022-01-28 08:26:01.406  INFO 1 --- [           main] o.s.nativex.NativeListener               : AOT mode disabled
        _                 _                                _
    ___(_)_ __ ___  _ __ | | ___       ___  ___ _ ____   _(_) ___ ___
   / __| | '_ ` _ \| '_ \| |/ _ \_____/ __|/ _ \ '__\ \ / / |/ __/ _ \
   \__ \ | | | | | | |_) | |  __/_____\__ \  __/ |   \ V /| | (_|  __/
   |___/_|_| |_| |_| .__/|_|\___|     |___/\___|_|    \_/ |_|\___\___|
                   |_|
   :: Spring Boot ::  (v2.6.3)
  
  2022-01-28 08:26:01.557  INFO 1 --- [           main] c.m.s.SimpleServiceApplication           : Starting SimpleServiceApplication using Java 11.0.13 on 6221adfb8770 with PID 1 (/app/classes started by root in /)
  2022-01-28 08:26:01.558  INFO 1 --- [           main] c.m.s.SimpleServiceApplication           : No active profile set, falling back to default profiles: default
  2022-01-28 08:26:03.408  INFO 1 --- [           main] o.s.b.w.embedded.tomcat.TomcatWebServer  : Tomcat initialized with port(s): 8080 (http)
  2022-01-28 08:26:03.424  INFO 1 --- [           main] o.apache.catalina.core.StandardService   : Starting service [Tomcat]
  2022-01-28 08:26:03.425  INFO 1 --- [           main] org.apache.catalina.core.StandardEngine  : Starting Servlet engine: [Apache Tomcat/9.0.56]
  2022-01-28 08:26:03.530  INFO 1 --- [           main] o.a.c.c.C.[Tomcat].[localhost].[/]       : Initializing Spring embedded WebApplicationContext
  2022-01-28 08:26:03.530  INFO 1 --- [           main] w.s.c.ServletWebServerApplicationContext : Root WebApplicationContext: initialization completed in 1864 ms
  2022-01-28 08:26:06.725  INFO 1 --- [           main] o.s.s.web.DefaultSecurityFilterChain     : Will secure any request with [org.springframework.security.web.context.request.async.WebAsyncManagerIntegrationFilter@2c8a445b, org.springframework.security.web.context.SecurityContextPersistenceFilter@5be51aa, org.springframework.security.web.header.HeaderWriterFilter@70f4f89e, org.springframework.security.web.csrf.CsrfFilter@1d283d1, org.springframework.security.web.authentication.logout.LogoutFilter@2de3b052, org.springframework.security.oauth2.client.web.OAuth2AuthorizationRequestRedirectFilter@892af0e, org.springframework.security.oauth2.client.web.OAuth2LoginAuthenticationFilter@446c8c72, org.springframework.security.web.authentication.ui.DefaultLoginPageGeneratingFilter@21be14c2, org.springframework.security.web.authentication.ui.DefaultLogoutPageGeneratingFilter@356b6b5d, org.springframework.security.web.savedrequest.RequestCacheAwareFilter@6c5747db, org.springframework.security.web.servletapi.SecurityContextHolderAwareRequestFilter@64c8fcfb, org.springframework.security.web.authentication.AnonymousAuthenticationFilter@37b1149b, org.springframework.security.web.session.SessionManagementFilter@74b1838, org.springframework.security.web.access.ExceptionTranslationFilter@2ba9f986, org.springframework.security.web.access.intercept.FilterSecurityInterceptor@50778bde]
  2022-01-28 08:26:06.754  INFO 1 --- [           main] o.s.b.a.e.web.EndpointLinksResolver      : Exposing 6 endpoint(s) beneath base path '/actuator'
  2022-01-28 08:26:06.810  INFO 1 --- [           main] o.s.b.w.embedded.tomcat.TomcatWebServer  : Tomcat started on port(s): 8080 (http) with context path ''
  2022-01-28 08:26:06.831  INFO 1 --- [           main] c.m.s.SimpleServiceApplication           : Started SimpleServiceApplication in 6.027 seconds (JVM running for 6.849)
  ```

---

- Running Native docker image with AOT mode enabled
  ```
  2022-01-28 08:40:37.515  INFO 1 --- [           main] o.s.nativex.NativeListener               : AOT mode enabled
  
    .   ____          _            __ _ _
   /\\ / ___'_ __ _ _(_)_ __  __ _ \ \ \ \
  ( ( )\___ | '_ | '_| | '_ \/ _` | \ \ \ \
   \\/  ___)| |_)| | | | | || (_| |  ) ) ) )
    '  |____| .__|_| |_|_| |_\__, | / / / /
   =========|_|==============|___/=/_/_/_/
   :: Spring Boot ::                (v2.6.3)
  
  2022-01-28 08:40:37.519  INFO 1 --- [           main] c.m.s.SimpleServiceApplication           : Starting SimpleServiceApplication using Java 11.0.13 on fd30bc8eec96 with PID 1 (/workspace/com.mycompany.simpleservice.SimpleServiceApplication started by cnb in /workspace)
  2022-01-28 08:40:37.519  INFO 1 --- [           main] c.m.s.SimpleServiceApplication           : No active profile set, falling back to default profiles: default
  2022-01-28 08:40:37.619  INFO 1 --- [           main] o.s.b.w.embedded.tomcat.TomcatWebServer  : Tomcat initialized with port(s): 8080 (http)
  2022-01-28 08:40:37.620  INFO 1 --- [           main] o.apache.catalina.core.StandardService   : Starting service [Tomcat]
  2022-01-28 08:40:37.620  INFO 1 --- [           main] org.apache.catalina.core.StandardEngine  : Starting Servlet engine: [Apache Tomcat/9.0.56]
  2022-01-28 08:40:37.625  INFO 1 --- [           main] o.a.c.c.C.[Tomcat].[localhost].[/]       : Initializing Spring embedded WebApplicationContext
  2022-01-28 08:40:37.625  INFO 1 --- [           main] w.s.c.ServletWebServerApplicationContext : Root WebApplicationContext: initialization completed in 105 ms
  2022-01-28 08:40:37.650  WARN 1 --- [           main] i.m.c.i.binder.jvm.JvmGcMetrics          : GC notifications will not be available because MemoryPoolMXBeans are not provided by the JVM
  2022-01-28 08:40:37.753  INFO 1 --- [           main] .s.s.UserDetailsServiceAutoConfiguration :
  
  Using generated security password: e324d...
  
  2022-01-28 08:40:37.755  INFO 1 --- [           main] o.s.s.web.DefaultSecurityFilterChain     : Will secure any request with [org.springframework.security.web.context.request.async.WebAsyncManagerIntegrationFilter@54ebc9d8, org.springframework.security.web.context.SecurityContextPersistenceFilter@8da06f4, org.springframework.security.web.header.HeaderWriterFilter@12643dd5, org.springframework.security.web.csrf.CsrfFilter@37fc8276, org.springframework.security.web.authentication.logout.LogoutFilter@36ecee93, org.springframework.security.web.savedrequest.RequestCacheAwareFilter@4bbfa6d7, org.springframework.security.web.servletapi.SecurityContextHolderAwareRequestFilter@267975a5, org.springframework.security.web.authentication.AnonymousAuthenticationFilter@38f77149, org.springframework.security.web.session.SessionManagementFilter@659220f, org.springframework.security.web.access.ExceptionTranslationFilter@e8020b9, org.springframework.security.web.access.intercept.FilterSecurityInterceptor@340f3fa4]
  2022-01-28 08:40:37.766  INFO 1 --- [           main] o.s.b.a.e.web.EndpointLinksResolver      : Exposing 6 endpoint(s) beneath base path '/actuator'
  2022-01-28 08:40:37.776  INFO 1 --- [           main] o.s.b.w.embedded.tomcat.TomcatWebServer  : Tomcat started on port(s): 8080 (http) with context path ''
  2022-01-28 08:40:37.777  INFO 1 --- [           main] c.m.s.SimpleServiceApplication           : Started SimpleServiceApplication in 0.276 seconds (JVM running for 0.279)
  ```

**Comment by snicoll** [comment]
Alright, thanks. At this point, this might be an Okta-specific problem that's doing something in the environment. 

@bdemers created https://github.com/spring-projects-experimental/spring-native/issues/1367 and I wonder if that's related since he's working on Okta.

**Comment by bdemers** [comment]
Hi @ivangfr, this does look like a duplicate of #1367
TL;DR - those Okta properties are _basically_ aliases of the equivalent Spring Sec props and it doesn't currently work with native-images.

Please feel free to create an issue on [okta/okta-spring-boot](https://github.com/okta/okta-spring-boot)

To work around the issue you should be able to use the following properties:

```properties
spring.security.oauth2.client.registration.okta.client-id=...
spring.security.oauth2.client.registration.okta.client-secret=...
spring.security.oauth2.client.provider.okta.issuer-uri=...
```

> **NOTE:** I do want to call out, I would not recommend baking sensitive data into your application bundle (jar, war, or executable). See the [External Configuration](https://docs.spring.io/spring-boot/docs/current/reference/html/features.html#features.external-config) for other options.

**Comment by ivangfr** [comment]
Thanks @bdemers for your help!

Sure, I will open an issue to okta/okta-spring-boot.

Btw, I am not backing the sensitive data into `application.properties`. That was just for testing. [I am setting them using environment variables](https://github.com/ivangfr/okta-springboot/blob/master/simple-service/src/main/resources/application.properties#L3).

Best regards

**Comment by ivangfr** [comment]
Hi @bdemers 

I am trying the properties you suggested as work around.

However, I am getting
```
WARN 81844 --- [           main] ironmentPostProcessorApplicationListener : Your Okta Issuer URL is missing.
You can copy your domain from the Okta Developer Console. Follow these instructions to find it: https://bit.ly/finding-okta-domain
To fix this add the `okta.oauth2.issuer` property to your application environments.
```

That is, message says the fix is to use `okta.oauth2.issuer`. So, the property `spring.security.oauth2.client.provider.okta.issuer-uri` seems not to work.

Thanks

**Comment by snicoll** [comment]
Closing as duplicate of #1367
